From 7ce5b04fa47253560d61b6afec5bd6373b3cb16a Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Sat, 2 Mar 2024 21:19:25 +0100
Subject: [PATCH 44/53] [HACK] mfmediaengine: Do not send
 MF_MEDIA_ENGINE_EVENT_ERROR for VRChat.

Breaks AVPro Video (used by VRChat) when a IMFMediaEngine::GetError call inside the event handler doesn't return an
IMFMediaError instance. The event handler code then tries to call a function on the non-existant IMFMediaError instance,
which segfaults.

I don't know why IMFMediaEngine::GetError returns a NULL instance inside the error callback. This should be fixed
properly. In the meantime, this hack works around the problem.

CW-Bug-Id: #21644
---
 dlls/mfmediaengine/main.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/dlls/mfmediaengine/main.c b/dlls/mfmediaengine/main.c
index 244e83a9f3e..c8d34cc4c82 100644
--- a/dlls/mfmediaengine/main.c
+++ b/dlls/mfmediaengine/main.c
@@ -1448,11 +1448,21 @@ static HRESULT WINAPI media_engine_load_handler_Invoke(IMFAsyncCallback *iface,
     }
     else
     {
-        engine->network_state = MF_MEDIA_ENGINE_NETWORK_NO_SOURCE;
-        engine->error_code = MF_MEDIA_ENGINE_ERR_SRC_NOT_SUPPORTED;
-        engine->extended_code = hr;
-        IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_ERROR, engine->error_code,
-                engine->extended_code);
+        static unsigned int once;
+        char sgi[64];
+
+        if (GetEnvironmentVariableA("SteamGameId", sgi, sizeof(sgi)) && !strcmp(sgi, "438100"))
+        {
+            if (!once++) FIXME("HACK: not sending MF_MEDIA_ENGINE_EVENT_ERROR event.\n");
+        }
+        else
+        {
+            engine->network_state = MF_MEDIA_ENGINE_NETWORK_NO_SOURCE;
+            engine->error_code = MF_MEDIA_ENGINE_ERR_SRC_NOT_SUPPORTED;
+            engine->extended_code = hr;
+            IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_ERROR, engine->error_code,
+                    engine->extended_code);
+        }
     }
 
     LeaveCriticalSection(&engine->cs);
-- 
2.47.1

